File: backend/internal/models/kanban.go
package models

import "time"

// Card priorities
const (
	CardPriorityBaixa   = "baixa"
	CardPriorityMedia   = "media"
	CardPriorityAlta    = "alta"
	CardPriorityUrgente = "urgente"
)

// Interaction types
const (
	InteractionComentario   = "comentario"
	InteractionSistemaMove  = "sistema_movimentacao"
	InteractionSistemaResp  = "sistema_responsavel"
	InteractionSistemaAnexo = "sistema_anexo"
	InteractionSistemaSupr  = "sistema_suprimentos"
)

type KanbanProject struct {
	ID                   uint      `gorm:"primaryKey" json:"id"`
	Titulo               string    `gorm:"size:255;not null" json:"titulo"`
	Descricao            *string   `gorm:"type:text" json:"descricao"`
	BoardBackgroundColor string    `gorm:"column:board_background_color;size:20;default:'#212121'" json:"board_background_color"`
	BoardPattern         string    `gorm:"column:board_pattern;size:30;default:'glow'" json:"board_pattern"`
	RelatedToMaintenance bool      `gorm:"column:related_to_maintenance;default:false" json:"related_to_maintenance"`
	RelatedToPreventive  bool      `gorm:"column:related_to_preventive;default:false" json:"related_to_preventive"`
	PreventivePlanID     *uint     `gorm:"column:preventive_plan_id" json:"preventive_plan_id"`
	CriadorID            uint      `gorm:"column:criador_id;not null" json:"criador_id"`
	IsActive             bool      `gorm:"column:is_active;default:true" json:"is_active"`
	IsArchived           bool      `gorm:"column:is_archived;default:false" json:"is_archived"`
	CreatedAt            time.Time `gorm:"column:created_at;default:CURRENT_TIMESTAMP" json:"created_at"`
	UpdatedAt            time.Time `gorm:"column:updated_at;default:CURRENT_TIMESTAMP" json:"updated_at"`

	Criador        *User            `gorm:"foreignKey:CriadorID" json:"criador,omitempty"`
	PreventivePlan *MaintenancePlan `gorm:"foreignKey:PreventivePlanID" json:"preventive_plan,omitempty"`
	Participantes  []User           `gorm:"many2many:kanban_project_participantes;" json:"participantes,omitempty"`
	Colunas        []KanbanColumn   `gorm:"foreignKey:ProjectID" json:"colunas,omitempty"`
	Cards          []KanbanCard     `gorm:"foreignKey:ProjectID" json:"cards,omitempty"`
}

func (KanbanProject) TableName() string { return "kanban_projects" }

type KanbanColumn struct {
	ID        uint   `gorm:"primaryKey" json:"id"`
	ProjectID uint   `gorm:"column:project_id;not null" json:"project_id"`
	Nome      string `gorm:"size:100;not null" json:"nome"`
	Cor       string `gorm:"size:30;default:'#6B7280'" json:"cor"`
	Ordem     int    `gorm:"default:0" json:"ordem"`
	IsDefault bool   `gorm:"column:is_default;default:false" json:"is_default"`

	Project *KanbanProject `gorm:"foreignKey:ProjectID" json:"project,omitempty"`
	Cards   []KanbanCard   `gorm:"foreignKey:ColumnID" json:"cards,omitempty"`
}

func (KanbanColumn) TableName() string { return "kanban_columns" }

type KanbanCard struct {
	ID                uint       `gorm:"primaryKey" json:"id"`
	ProjectID         uint       `gorm:"column:project_id;not null" json:"project_id"`
	ColumnID          uint       `gorm:"column:column_id;not null" json:"column_id"`
	Titulo            string     `gorm:"size:255;not null" json:"titulo"`
	Cor               string     `gorm:"column:cor;size:20;default:'#0079BF'" json:"cor"`
	Descricao         *string    `gorm:"type:text" json:"descricao"`
	ChecklistJSON     *string    `gorm:"column:checklist_json;type:text" json:"checklist_json"`
	PreventiveOrderID *uint      `gorm:"column:preventive_order_id" json:"preventive_order_id"`
	CriadorID         uint       `gorm:"column:criador_id;not null" json:"criador_id"`
	ResponsavelID     *uint      `gorm:"column:responsavel_id" json:"responsavel_id"`
	Prioridade        string     `gorm:"size:20;default:'media'" json:"prioridade"`
	DataEntrega       *time.Time `gorm:"column:data_entrega" json:"data_entrega"`
	Ordem             int        `gorm:"default:0" json:"ordem"`
	CreatedAt         time.Time  `gorm:"column:created_at;default:CURRENT_TIMESTAMP" json:"created_at"`
	UpdatedAt         time.Time  `gorm:"column:updated_at;default:CURRENT_TIMESTAMP" json:"updated_at"`

	// Procurement links (plain columns; FK constraints are managed by the procurement module)
	PurchaseRequestID  *uint   `gorm:"column:purchase_request_id" json:"purchase_request_id"`
	MaterialStockID    *uint   `gorm:"column:material_stock_id" json:"material_stock_id"`
	TipoItemNecessario *string `gorm:"column:tipo_item_necessario;size:50" json:"tipo_item_necessario"`

	Project         *KanbanProject          `gorm:"foreignKey:ProjectID" json:"project,omitempty"`
	Column          *KanbanColumn           `gorm:"foreignKey:ColumnID" json:"column,omitempty"`
	PreventiveOrder *MaintenanceOrder       `gorm:"foreignKey:PreventiveOrderID" json:"preventive_order,omitempty"`
	Criador         *User                   `gorm:"foreignKey:CriadorID" json:"criador,omitempty"`
	Responsavel     *User                   `gorm:"foreignKey:ResponsavelID" json:"responsavel,omitempty"`
	Participantes   []User                  `gorm:"many2many:kanban_card_participantes;" json:"participantes,omitempty"`
	Ativos          []Asset                 `gorm:"many2many:kanban_card_assets;" json:"ativos,omitempty"`
	Anexos          []KanbanAttachment      `gorm:"foreignKey:CardID" json:"anexos,omitempty"`
	Interacoes      []KanbanCardInteraction `gorm:"foreignKey:CardID" json:"interacoes,omitempty"`
}

func (KanbanCard) TableName() string { return "kanban_cards" }

type KanbanCardInteraction struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	CardID    uint      `gorm:"column:card_id;not null;index" json:"card_id"`
	UsuarioID uint      `gorm:"column:usuario_id;not null" json:"usuario_id"`
	Mensagem  string    `gorm:"type:text;not null" json:"mensagem"`
	Tipo      string    `gorm:"size:50;default:'comentario'" json:"tipo"`
	CreatedAt time.Time `gorm:"column:created_at;default:CURRENT_TIMESTAMP" json:"created_at"`

	Card    *KanbanCard `gorm:"foreignKey:CardID" json:"card,omitempty"`
	Usuario *User       `gorm:"foreignKey:UsuarioID" json:"usuario,omitempty"`
}

func (KanbanCardInteraction) TableName() string { return "kanban_card_interactions" }

type KanbanAttachment struct {
	ID       uint      `gorm:"primaryKey" json:"id"`
	CardID   uint      `gorm:"column:card_id;not null" json:"card_id"`
	Nome     string    `gorm:"size:255;not null" json:"nome"`
	Tipo     string    `gorm:"size:50;not null" json:"tipo"` // imagem, link, arquivo
	URL      string    `gorm:"type:text;not null" json:"url"`
	CriadoEm time.Time `gorm:"column:criado_em;default:CURRENT_TIMESTAMP" json:"criado_em"`

	Card *KanbanCard `gorm:"foreignKey:CardID" json:"card,omitempty"`
}

func (KanbanAttachment) TableName() string { return "kanban_attachments" }

// Notification types
const (
	NotifProjetoAdicionado = "PROJETO_ADICIONADO"
	NotifCartaoAtribuido   = "CARTAO_ATRIBUIDO"
	NotifCartaoMovimentado = "CARTAO_MOVIMENTADO"
	NotifAnexoAdicionado   = "ANEXO_ADICIONADO"
)

type KanbanNotification struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	UserID    uint      `gorm:"column:user_id;not null;index" json:"user_id"`
	ProjectID *uint     `gorm:"column:project_id" json:"project_id"`
	CardID    *uint     `gorm:"column:card_id" json:"card_id"`
	AutorID   *uint     `gorm:"column:autor_id" json:"autor_id"`
	Tipo      string    `gorm:"size:50;not null" json:"tipo"`
	Titulo    string    `gorm:"size:255;not null" json:"titulo"`
	Mensagem  string    `gorm:"type:text;not null" json:"mensagem"`
	Link      *string   `gorm:"size:255" json:"link"`
	Lida      bool      `gorm:"default:false" json:"lida"`
	CreatedAt time.Time `gorm:"column:created_at;default:CURRENT_TIMESTAMP" json:"created_at"`

	User  *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Autor *User `gorm:"foreignKey:AutorID" json:"autor,omitempty"`
}

func (KanbanNotification) TableName() string { return "kanban_notifications" }

File: frontend/src/types/kanban.ts
export interface KanbanProject {
  id: number;
  titulo: string;
  descricao?: string;
  board_background_color?: string;
  board_pattern?: 'glow' | 'grid' | 'dots' | 'clean';
  related_to_maintenance?: boolean;
  related_to_preventive?: boolean;
  preventive_plan_id?: number;
  preventive_automation_enabled?: boolean;
  preventive_automation_horizon_days?: number;
  preventive_automation_started_at?: string | null;
  preventive_automation_last_run_at?: string | null;
  preventive_automation_next_run_at?: string | null;
  preventive_card_title_template?: string;
  preventive_card_description_template?: string;
  preventive_card_checklist_template?: string;
  preventive_card_priority?: string;
  preventive_card_color?: string;
  criador_id: number;
  is_active: boolean;
  is_archived: boolean;
  favoritado?: boolean;
  created_at: string;
  updated_at: string;
  criador?: { id: number; nome: string };
  preventive_plan?: { id: number; nome: string; codigo: string };
  participantes?: { id: number; nome: string; avatar_url?: string | null }[];
  colunas?: KanbanColumn[];
  cards?: KanbanCard[];
}

export interface KanbanColumn {
  id: number;
  project_id: number;
  nome: string;
  cor: string;
  ordem: number;
  is_default: boolean;
  cards?: KanbanCard[];
}

export interface KanbanChecklistItem {
  id: string;
  titulo: string;
  concluido: boolean;
}

export interface KanbanCard {
  id: number;
  project_id: number;
  column_id: number;
  titulo: string;
  cor?: string;
  descricao?: string;
  checklist_json?: string;
  checklist_items?: KanbanChecklistItem[];
  preventive_order_id?: number;
  criador_id: number;
  responsavel_id?: number;
  prioridade: string;
  data_entrega?: string;
  ordem: number;
  created_at: string;
  updated_at: string;
  purchase_request_id?: number;
  material_stock_id?: number;
  tipo_item_necessario?: string;
  column?: KanbanColumn;
  criador?: { id: number; nome: string; avatar_url?: string | null };
  responsavel?: { id: number; nome: string; avatar_url?: string | null };
  preventive_order?: {
    id: number;
    numero: string;
    status: string;
    prioridade: string;
    tipo: string;
    data_agendada?: string;
    asset?: { id: number; nome: string; e_patrimonio: string };
    tecnico?: { id: number; nome: string };
    plan?: { id: number; nome: string; codigo: string };
  };
  participantes?: { id: number; nome: string; avatar_url?: string | null }[];
  ativos?: { id: number; nome: string }[];
  anexos?: KanbanAttachment[];
  interacoes?: KanbanCardInteraction[];
}

export interface KanbanCardInteraction {
  id: number;
  card_id: number;
  usuario_id: number;
  mensagem: string;
  tipo: string;
  created_at: string;
  usuario?: { id: number; nome: string };
}

export interface KanbanAttachment {
  id: number;
  card_id: number;
  nome: string;
  tipo: string; // imagem, link, arquivo
  url: string;
  criado_em: string;
}

export interface KanbanNotification {
  id: number;
  user_id: number;
  project_id?: number;
  card_id?: number;
  autor_id?: number;
  tipo: string;
  titulo: string;
  mensagem: string;
  link?: string;
  lida: boolean;
  created_at: string;
  autor?: { id: number; nome: string };
}

export const CARD_PRIORITIES = ['baixa', 'media', 'alta', 'urgente'];

export const priorityColor: Record<string, string> = {
  baixa: 'text-brand-muted border-brand-border',
  media: 'text-blue-400 border-blue-500/30',
  alta: 'text-yellow-400 border-yellow-500/30',
  urgente: 'text-red-400 border-red-500/30',
};
